src/NHibernate/Persister/Entity/AbstractEntityPersister.cs

```diff
@@ -2456,6 +2454,10 @@ protected int Dehydrate(object id, object[] fields, bool[] includeProperty, bool
 						PropertyTypes[i].NullSafeSet(statement, fields[i], index, includeColumns[i], session);
 						index += ArrayHelper.CountTrue(includeColumns[i]); //TODO:  this is kinda slow...
 					}
+					catch (ArgumentOutOfRangeException arex)
```

Comment:
It seems the exception is an `IndexOutOfRangeException`, not an `ArgumentOutOfRangeException`.

Anyway, it would be better to avoid the `IndexOutOfRangeException` by detecting the root trouble sooner. It should be doable when this class is constructed, or maybe even sooner. Then the exception would likely be thrown when compiling mappings, which is better.

Reply:
I'm looking at the error right now, it says:

> NHibernate.PropertyValueException : Error dehydrating property value for PatientService.Logic.Persistence.Models.PersistentPatient.EncryptedObsolete
>   ----> System.ArgumentOutOfRangeException : Index was out of range. Must be non-negative and less than the size of the collection.
> Parameter name: index

I agree, this should break when compiling mappings. That's a change I'm not comfortable with, though. I _can_ improve the error message :-)

Reply:
The error in TeamCity for SQL-Server is an `IndexOutOfRangeException`, see [here](https://teamcity.jetbrains.com/viewLog.html?buildId=1685546&buildTypeId=bt7) (use anonymous login option if you do not have a login). With NHibernate Teamcity configuration, all other databases builds depend on the "NHibernate (NHibernate)" build, and are not actually executed if this first one has failed, so we cannot check the error for other databases tested with Teamcity.
But on this previous appveyor [build](https://ci.appveyor.com/project/nhibernate/nhibernate-core/builds/19552572) for your PR, we can see the error is not the same depending on the target database:
- [PostgreSQL](https://ci.appveyor.com/project/nhibernate/nhibernate-core/builds/19552572/job/a9wubk2f86rjjo2p/tests): `ArgumentOutOfRangeException`.
- [Firebird](https://ci.appveyor.com/project/nhibernate/nhibernate-core/builds/19552572/job/xe92qy8gumnmo5ih/tests) apparently no exception but failure of asserts (which one, hard to say, since that were the previous assert, just checking boolean results).
- [MySql](https://ci.appveyor.com/project/nhibernate/nhibernate-core/builds/19552572/job/3fg1pe8a40xcx50c/tests): `IndexOutOfRangeException`.
- [SQLite](https://ci.appveyor.com/project/nhibernate/nhibernate-core/builds/19552572/job/etgwegxmlgwx0oj4/tests): `ArgumentOutOfRangeException`.
- [SQL-Server](https://ci.appveyor.com/project/nhibernate/nhibernate-core/builds/19552572/job/cnreul4cj6cj4cjq/tests): `IndexOutOfRangeException`.

So it is now clear that catching the exception and trying to translate it to a more meaningful error message is not appropriate: it will not be reliable, since the thrown exception depends on the target database.

By the way, all the links above are provided by GitHub on this PR. On a commit, click the status icon (red cross or green check) to see the list of builds, and click on their "Details" link. Or alternatively for the last commit, see the checks panel at the bottom of the PR.